icicle/tests/test_hash_api.cpp

```diff
@@ -313,38 +316,210 @@ class HashSumBackend : public HashBackend
   }
 };
 
-TEST_F(HashApiTest, MerkleTree)
+void assert_valid_tree(
+  const MerkleTree& tree,
+  int nof_input_byes,
+  const std::byte* inputs,
+  const std::vector<Hash>& hashes,
+  const MerkleTreeConfig& config)
+{
+  int nof_outputs = nof_input_byes * hashes[0].output_size() / hashes[0].input_default_chunk_size();
+  std::byte* layer_in = new std::byte[nof_input_byes]; // Going layer by layer - having the input layer as the largest
+  std::byte* layer_out = new std::byte[nof_outputs];   // ensures these are the maximum sizes required for the arrays
+  // NOTE there is an assumption here that output number is less or equal to input number for all layers
+
+  for (int i = 0; i < nof_input_byes; i++) {
+    layer_in[i] = inputs[i];
+  }
+
+  // int nof_hashes = nof_inputs / hashes[0]->m_total_input_limbs;
+  int side_inputs_offset = 0;
+  for (int i = 0; i < hashes.size(); i++) {
+    nof_outputs = nof_input_byes * hashes[i].output_size() / hashes[i].input_default_chunk_size();
+    const int nof_hashes = nof_input_byes / hashes[i].input_default_chunk_size();
+
+    auto config = default_hash_config();
+    config.batch = nof_hashes;
+    hashes[i].hash(layer_in, hashes[i].input_default_chunk_size(), config, layer_out);
+
+    // Transfer outputs to inputs before moving to the next layer
+    for (int i = 0; i < nof_outputs; i++) {
+      layer_in[i] = layer_out[i];
+    }
+    nof_input_byes = nof_outputs;
+  }
+
+  // Compare computed root with the tree's root
+  auto [root, root_size] = tree.get_merkle_root();
+
+  for (int i = 0; i < root_size; i++) {
+    ICICLE_ASSERT(root[i] == layer_out[i]) << "Mismatch in root[" << i << "]";
+  }
+
+  delete[] layer_in;
+  delete[] layer_out;
+}
+
+template <typename in_limb>
+void assert_valid_tree(
+  const MerkleTree& tree,
+  int nof_inputs,
+  const in_limb* inputs,
+  const std::vector<Hash>& hashes,
+  const MerkleTreeConfig& config)
+{
+  return assert_valid_tree(
```

Comment:
return type is void so no need the return statement.
Maybe better to return a bool and not assert though.